### ISSUE TITLE
python3Packages: get rid of uses of nose

### ISSUE THREAD

**Issue by emilazy** [NixOS/nixpkgs#326513]
nose is [abandoned upstream](https://nose.readthedocs.io/en/latest/) and has caused us untold headaches recently (admittedly mostly unrelated to the code itself). We’ve stopped the bleeding by getting https://github.com/NixOS/nixpkgs/pull/325968 into staging thanks to the incredible work of @jchv, but it’s still broken on master and we ought to do some tidying up so we’re not dragging this around forever and perhaps even to unblock some stuff on master while the staging cycle progresses. Many packages don’t actually need nose any more, despite us including it in our dependency lists. Active upstreams should have already moved to pytest or something else. Some packages will have patches available; if we’re really feeling generous we could send patches to upstream ourselves. Others may be completely unmaintained packages that we could think about dropping.

@dotlambda has already done some great work on this – see https://github.com/NixOS/nixpkgs/pull/325847, https://github.com/NixOS/nixpkgs/pull/325872, https://github.com/NixOS/nixpkgs/pull/325885, and https://github.com/NixOS/nixpkgs/pull/326164 – but I thought it’d be good to have a master list of all affected packages so we can tackle it collectively.

* [ ] `adafruit-nrfutil`
* [ ] `aws-google-auth`
* [ ] `caffe`
* [ ] `candidates1`
* [ ] `cmake-lint`
* [ ] `gixy`
* [ ] `hydrus`
* [ ] `ledger-autosync`
* [ ] `mackup`
* [ ] `mbutil`
* [ ] `onboard`
* [ ] `openlp`
* [ ] `python3Packages.actdiag`
* [ ] `python3Packages.androguard`
* [ ] `python3Packages.apricot-select`
* [ ] `python3Packages.beaker`
* [ ] `python3Packages.biopandas`
* [ ] `python3Packages.blessings`
* [ ] `python3Packages.blockdiag`
* [ ] `python3Packages.boto`
* [ ] `python3Packages.braintree`
* [ ] `python3Packages.bx-python`
* [ ] `python3Packages.cassandra-driver`
* [ ] `python3Packages.cgroup-utils`
* [ ] `python3Packages.changefinder`
* [ ] `python3Packages.chart-studio`
* [ ] `python3Packages.citeproc-py`
* [ ] `python3Packages.clickhouse-driver`
* [ ] `python3Packages.clustershell`
* [ ] `python3Packages.deform`
* [ ] `python3Packages.django-crontab`
* [ ] `python3Packages.dm-control`
* [ ] `python3Packages.docker`
* [ ] `python3Packages.dodgy`
* [ ] `python3Packages.envs`
* [ ] `python3Packages.expiringdict`
* [ ] `python3Packages.flask-restful`
* [ ] `python3Packages.gocardless-pro`
* [ ] `python3Packages.habanero`
* [ ] `python3Packages.hkdf`
* [ ] `python3Packages.http-ece`
* [ ] `python3Packages.inotify`
* [ ] `python3Packages.isbnlib`
* [ ] `python3Packages.jsonable`
* [ ] `python3Packages.jupyter-contrib-nbextensions`
* [ ] `python3Packages.libgpuarray`
* [ ] `python3Packages.locationsharinglib`
* [ ] `python3Packages.lockfile`
* [ ] `python3Packages.mhcflurry`
* [ ] `python3Packages.mkl-service`
* [ ] `python3Packages.mohawk`
* [ ] `python3Packages.mongoengine`
* [ ] `python3Packages.more-itertools`
* [ ] `python3Packages.mwtypes`
* [ ] `python3Packages.mwxml`
* [ ] `python3Packages.mygpoclient`
* [ ] `python3Packages.nbmerge`
* [ ] `python3Packages.neo`
* [ ] `python3Packages.nose-pattern-exclude`
* [ ] `python3Packages.nose-timer`
* [ ] `python3Packages.nose-warnings-filters`
* [ ] `python3Packages.nose-xunitmp`
* [ ] `python3Packages.nwdiag`
* [ ] `python3Packages.ofxhome`
* [ ] `python3Packages.ofxtools`
* [ ] `python3Packages.para`
* [ ] `python3Packages.paramz`
* [ ] `python3Packages.pastescript`
* [ ] `python3Packages.paver`
* [ ] `python3Packages.pid`
* [ ] `python3Packages.pkutils`
* [ ] `python3Packages.pprintpp`
* [ ] `python3Packages.preggy`
* [ ] `python3Packages.premailer`
* [ ] `python3Packages.prison`
* [ ] `python3Packages.prox-tv`
* [ ] `python3Packages.ptable`
* [ ] `python3Packages.pycdio`
* [ ] `python3Packages.pycontracts`
* [ ] `python3Packages.pydy`
* [ ] `python3Packages.pyexcel-ods`
* [ ] `python3Packages.pyexcel-xls`
* [ ] `python3Packages.pygatt`
* [ ] `python3Packages.pygeoip`
* [ ] `python3Packages.pygogo`
* [ ] `python3Packages.pygtfs`
* [ ] `python3Packages.pylacrosse`
* [ ] `python3Packages.pypass`
* [ ] `python3Packages.pypillowfight`
* [ ] `python3Packages.pyquaternion`
* [ ] `python3Packages.pyrfc3339`
* [ ] `python3Packages.pysrt`
* [ ] `python3Packages.pytest`
* [ ] `python3Packages.python-etcd`
* [ ] `python3Packages.python-fedora`
* [ ] `python3Packages.python-hglib`
* [ ] `python3Packages.python-mapnik`
* [ ] `python3Packages.pytimeparse`
* [ ] `python3Packages.pyutilib`
* [ ] `python3Packages.rauth`
* [ ] `python3Packages.rx`
* [ ] `python3Packages.safe`
* [ ] `python3Packages.sampledata`
* [ ] `python3Packages.scales`
* [ ] `python3Packages.scikit-fuzzy`
* [ ] `python3Packages.selectors2`
* [ ] `python3Packages.seqdiag`
* [ ] `python3Packages.shap`
* [ ] `python3Packages.simplebayes`
* [ ] `python3Packages.spark-parser`
* [ ] `python3Packages.sphinx-rtd-dark-mode`
* [ ] `python3Packages.sqlalchemy-mixins`
* [ ] `python3Packages.stytra`
* [ ] `python3Packages.sure`
* [ ] `python3Packages.svgutils`
* [ ] `python3Packages.telfhash`
* [ ] `python3Packages.tempita`
* [ ] `python3Packages.tissue`
* [ ] `python3Packages.traittypes`
* [ ] `python3Packages.trfl`
* [ ] `python3Packages.uncompyle6`
* [ ] `python3Packages.unicode-slugify`
* [ ] `python3Packages.uvcclient`
* [ ] `python3Packages.vxi11`
* [ ] `python3Packages.webassets`
* [ ] `python3Packages.webhelpers`
* [ ] `python3Packages.weboob`
* [ ] `python3Packages.wsme`
* [ ] `python3Packages.xlwt`
* [ ] `python3Packages.yanc`
* [ ] `python3Packages.yarg`
* [ ] `sasview`
* [ ] `svtplay-dl`
* [ ] `tarsnapper`
* [ ] `zk-shell`

This was generated from staging commit 83881cb6e9b660cbca5241380c031ab2b69d6e1e with the following awful fish one‐liner: `for pkg in (rg -l '\bnose3?\b' | rg -v '/(nose[23]?|cufflinks|enocean|jaconv|annoy|hdmedians|flaky|pycron|snapshottest|grin|ssh-mitm|colour|influxdb)/|(python|cran)-(packages|aliases)\.nix' | sed 's|.*/\([^/]*\)/\([^/]*\)$|\1|'); rg "/$pkg"'[ \n{]' pkgs/top-level/python-packages.nix | awk '{print "python3Packages." $1}'; test $pipestatus[1] = 0 || rg "/$pkg"'[ \n{]' pkgs/top-level/all-packages.nix | awk '{print $1}'; test $pipestatus[1] = 0 || echo "$pkg"; end | sort -u | sed 's/^/* [ ] `/; s/$/`/'`; feel free to edit to update the list.

**Comment by lavafroth** [comment]
`python3Packages.biopandas` has [dropped pynose in v0.5.1.dev1](https://github.com/BioPandas/biopandas/pull/146) (not yet released). We should be able to check that off once the next version releases.

**Comment by cyrinux** [comment]
Hi, "paperwork" package  also depend on nose and raise error actually.

**Comment by pyrox0** [comment]
Seems like paperwork depends on pypillowfight, which has a build error due to nose not being supported. I'm investigating now!

**Comment by pyrox0** [comment]
`ledger-autosync` can stay broken for now, per https://github.com/egh/ledger-autosync/issues/143, upstream is working on getting a new release out, which should include all fixes for distutils, nose -> pytest, etc. I could also just update to the latest commit in the meantime if that's what's wanted.

**Comment by pyrox0** [comment]
`sasview`, while listed here, does not depend on nose.

**Comment by lavafroth** [comment]
@pyrox0 `sasview` has an `xhtml2pdf.nix` besides the `default.nix` which depends on nose.

**Comment by Sigmanificient** [comment]
I am not a package maintainer of any of the mentioned packages, may I still help?

**Comment by pyrox0** [comment]
@Sigmanificient yes you may! feel free to submit a PR for any listed package ^^

**Comment by eclairevoyant** [comment]
Yep, anyone can submit a PR for any package, as long as it follows guidelines; maintainership != exclusive ownership.

**Comment by lavafroth** [comment]
We should probably mark the packages we have discovered that still use `nose` so that we don't have to repeat each other's homework.

**Comment by emilazy** [comment]
As in, the packages that would need patching to drop the dependency? Seems like a good idea; if anyone without permissions to edit the original post lists such packages here I’ll add a note to the list and hide the comment to avoid noise.

**Comment by Sigmanificient** [comment]
What should be done for `trfl`? It is marked as broken

**Comment by eclairevoyant** [comment]
Drop it? It hasn't seen activity in the 2 years since it was init'ed...

**Comment by Sigmanificient** [comment]
> Drop it? It hasn't seen activity in the 2 years since it was init'ed...

Oh this is reassuring, I was about to do a drop pr, abut I noticed that it was marked as broken only 2 month ago.

**Comment by Sigmanificient** [comment]
`styra` was marked as broken 8 months ago: https://github.com/NixOS/nixpkgs/commits/master/pkgs/development/python-modules/stytra/default.nix

**Comment by lavafroth** [comment]
[`pyrfc3339`](https://github.com/kurtraschke/pyRFC3339) needs patching to drop nose. Last commit was 5 years ago. Python 3.2+ appears to have support for this natively (see [here](https://stackoverflow.com/questions/8556398/generate-rfc-3339-timestamp-in-python#8556555)). Should we drop it?

**Comment by emilazy** [comment]
Yes, let’s drop it. In general, anything that’s giving us trouble during the 3.12 migration and clearly has no maintenance is most likely going to cause us more trouble in future; the same reasons we want to get rid of nose apply to most of the packages that will make getting rid of nose a pain.

Of course, there are some packages that may be too popular or too irreplaceable for dropping them to be easy to stomach, but I do think that in the cases where upstream actually does depend on nose and doesn’t seem actively maintained enough for a PR to be worthwhile, we should lean towards just dropping the package by default. Those packages are going to fall out of the active Python ecosystem anyway because you won’t be able to use them on the latest Python version with the standard package managers; in the absence of a compelling reason to make an exception, it doesn’t do much good for us to keep them on life support only for users of Nixpkgs.

It’s easy to re‐add a removed package if people complain or if something that requires it gets added to Nixpkgs, but hard for us to actively keep broken packages working. We should think of this as an opportunity to tend our garden rather than just doing a bunch of labour for dead upstreams and no real benefit. As far as making the judgement call goes, in‐tree reverse dependencies are an obvious metric to use for libraries, but I also found a site called [Wheelodex](https://www.wheelodex.org/) that indexes reverse dependencies for PyPI projects. That can help inform the judgement calls about whether to patch or drop.

**Comment by eclairevoyant** [comment]
Do you know an easy way to build an in-tree reverse dependency list for a given package?

**Comment by emilazy** [comment]
> Do you know an easy way to build an in-tree reverse dependency list for a given package? (As in the full tree, not just one level up)

No, I’ve just been grepping for immediate reverse dependencies and then repeating if anything looks likely to be used by anything else. I know there’s [a script](https://discourse.nixos.org/t/how-do-you-find-reverse-dependencies/15057/7) by @trofi that probably does a better job, and ofborg of course knows how to do it.

By the way, in the case of pyRFC3339, it looks like Certbot depends on it. Maybe that’s just a compatibility thing and it can use the standard library on newer versions, but if not we probably can’t drop it.

**Comment by emilazy** [comment]
It calculates the number of rebuilds to tag by checking which output paths change. The ofborg-eval action output lists all of them.

**Comment by emilazy** [comment]
Maybe this will come in handy for stuff we need to patch: https://github.com/pytest-dev/nose2pytest.

**Comment by eclairevoyant** [comment]
> It calculates the number of rebuilds to tag by checking which output paths change. The ofborg-eval action output lists all of them.

yes but it gets the number of changed hydraJobs, it doesn't query for reverse deps (afaik)

**Comment by trofi** [comment]
Using https://github.com/trofi/nixpkgs-overlays/blob/main/lib/arevdeps.nix I got this list:

<details>
<summary>$ nix-instantiate --impure --read-write-mode --eval --strict ~/.config/nixpkgs/lib/arevdeps.nix --argstr attrName python311Packages.nose -I nixpkgs=$PWD --arg verbose 2 --arg depth 2</summary>

```
python311Packages.actdiag
python311Packages.biopandas
python311Packages.blockdiag
python311Packages.boto
python311Packages.braintree
python311Packages.bx-python
python311Packages.cassandra-driver
python311Packages.cgroup-utils
python311Packages.changefinder
python311Packages.chart-studio
python311Packages.dm-control
python311Packages.dodgy
python311Packages.envs
python311Packages.flask-restful
python311Packages.hkdf
python311Packages.http-ece
python311Packages.jsonable
python311Packages.libgpuarray
python311Packages.locationsharinglib
python311Packages.lockfile
python311Packages.mohawk
python311Packages.mwtypes
python311Packages.mwxml
python311Packages.neo
python311Packages.nose
python311Packages.nose-pattern-exclude
python311Packages.nose-timer
python311Packages.nose-warnings-filters
python311Packages.nose-xunitmp
python311Packages.nwdiag
python311Packages.ofxhome
python311Packages.ofxtools
python311Packages.para
python311Packages.paramz
python311Packages.paver
python311Packages.pid
python311Packages.pprintpp
python311Packages.preggy
python311Packages.premailer
python311Packages.prison
python311Packages.prox-tv
python311Packages.ptable
python311Packages.pycdio
python311Packages.pycontracts
python311Packages.pydy
python311Packages.pyexcel-ods
python311Packages.pyexcel-xls
python311Packages.pygatt
python311Packages.pygeoip
python311Packages.pygogo
python311Packages.pygtfs
python311Packages.pylacrosse
python311Packages.pypass
python311Packages.pyquaternion
python311Packages.pyrfc3339
python311Packages.pysrt
python311Packages.python-etcd
python311Packages.python-hglib
python311Packages.python-mapnik
python311Packages.pytimeparse
python311Packages.pyutilib
python311Packages.rauth
python311Packages.safe
python311Packages.sampledata
python311Packages.scikit-fuzzy
python311Packages.selectors2
python311Packages.seqdiag
python311Packages.sphinx-rtd-dark-mode
python311Packages.sqlalchemy-mixins
python311Packages.sure
python311Packages.svgutils
python311Packages.telfhash
python311Packages.tempita
python311Packages.tissue
python311Packages.traittypes
python311Packages.unicode-slugify
python311Packages.vxi11
python311Packages.webhelpers
python311Packages.weboob
python311Packages.xlwt
```
</details>

**Comment by Sigmanificient** [comment]
I'll try to keep a checklist up to date, based on @trofi's list:

- [ ] python311Packages.actdiag
- [ ] python311Packages.biopandas
- [ ] python311Packages.blockdiag
- [ ] python311Packages.boto
- [X] python311Packages.braintree | #330290
- [ ] python311Packages.bx-python
- [ ] python311Packages.cassandra-driver
- [X] python311Packages.citeproc-py | #330237
- [ ] python311Packages.cgroup-utils | #330287
- [ ] python311Packages.changefinder
- [ ] python311Packages.chart-studio | #330248
- [ ] python311Packages.dm-control
- [ ] python311Packages.dodgy
- [ ] python311Packages.envs
- [ ] python311Packages.flask-restful
- [ ] python311Packages.hkdf
- [ ] python311Packages.http-ece
- [ ] python311Packages.jsonable
- [ ] python311Packages.libgpuarray
- [ ] python311Packages.locationsharinglib
- [ ] python311Packages.lockfile
- [ ] python311Packages.mohawk
- [ ] python311Packages.mwtypes
- [ ] python311Packages.mwxml
- [ ] python311Packages.neo
- [ ] python311Packages.nose
- [ ] python311Packages.nose-pattern-exclude
- [ ] python311Packages.nose-timer
- [ ] python311Packages.nose-warnings-filters
- [ ] python311Packages.nose-xunitmp
- [ ] python311Packages.nwdiag
- [ ] python311Packages.ofxhome
- [ ] python311Packages.ofxtools
- [ ] python311Packages.para | #330229
- [ ] python311Packages.paramz
- [ ] python311Packages.paver
- [ ] python311Packages.pid
- [ ] python311Packages.pprintpp
- [ ] python311Packages.preggy
- [ ] python311Packages.premailer
- [ ] python311Packages.prison
- [ ] python311Packages.prox-tv
- [ ] python311Packages.ptable
- [ ] python311Packages.pycdio
- [ ] python311Packages.pycontracts
- [ ] python311Packages.pydy
- [ ] python311Packages.pyexcel-ods
- [ ] python311Packages.pyexcel-xls
- [ ] python311Packages.pygatt
- [ ] python311Packages.pygeoip
- [ ] python311Packages.pygogo
- [ ] python311Packages.pygtfs
- [ ] python311Packages.pylacrosse
- [ ] python311Packages.pypass
- [ ] python311Packages.pyquaternion
- [ ] python311Packages.pyrfc3339
- [ ] python311Packages.pysrt
- [ ] python311Packages.python-etcd
- [ ] python311Packages.python-hglib
- [ ] python311Packages.python-mapnik
- [ ] python311Packages.pytimeparse
- [ ] python311Packages.pyutilib
- [ ] python311Packages.rauth
- [ ] python311Packages.safe
- [ ] python311Packages.sampledata
- [ ] python311Packages.scikit-fuzzy
- [ ] python311Packages.selectors2
- [ ] python311Packages.seqdiag
- [ ] python311Packages.sphinx-rtd-dark-mode
- [ ] python311Packages.sqlalchemy-mixins
- [ ] python311Packages.sure
- [ ] python311Packages.svgutils
- [ ] python311Packages.telfhash
- [ ] python311Packages.tempita
- [ ] python311Packages.tissue
- [ ] python311Packages.traittypes
- [ ] python311Packages.unicode-slugify
- [ ] python311Packages.vxi11
- [ ] python311Packages.webhelpers
- [ ] python311Packages.weboob
- [ ] python311Packages.xlwt

**Comment by emilazy** [comment]
The list in my original issue is the relevant one. The packages present in @trofi’s list but not mine don’t depend directly on nose, so there’s nothing actionable to do with them. The reverse dependencies script is potentially useful when assessing whether to drop individual packages in my list.

**Comment by emilazy** [comment]
Ah, sorry; I forgot that `arevdeps.nix` only does one level of reverse dependencies. So actually it’s relevant, and maybe some of the differences between the two might be due to the imprecision of my grep script.

**Comment by Sigmanificient** [comment]
[Boto](https://github.com/boto/boto) was archived 10 may 2024

**Comment by emilazy** [comment]
Although boto is deprecated, a few things still depend on it, probably including out‐of‐tree projects. I’m not qualified to say if it’d be possible to do an immediate sweep of them, but I’d probably lean towards fixing it instead.

**Comment by lavafroth** [comment]
Seems like package weboob is a duplicate of woob?

**Comment by Sigmanificient** [comment]
webhelpers has `doCheck = false` due to broken code, but I cant access the source
Is is on my side, or repository is gone: https://bitbucket.org/bbangert/webhelpers

**Comment by lavafroth** [comment]
@Sigmanificient the source code should still be there on Pypi but yes, the original repo is gone.

**Comment by Sigmanificient** [comment]
I think it should be dropped, as it was not updated (release) from when it as been mode to `pkgs/development/python-modules` 6 years ago.
It dont have any reverse dependency

**Comment by emilazy** [comment]
> Seems like package weboob is a duplicate of woob?

Uhh, wow. The things you can find hiding in the corners of a repository this large. The one with a maintainer seems more polished and has a newer version and more current licence. But wow. We should definitely drop `weboob`; looks like upstream has used pytest since 2021, so it should be a pretty easy fix for the other one.

**Comment by lavafroth** [comment]
[pylockfile](http://git.openstack.org/cgit/openstack/pylockfile) is no longer maintained.

**Comment by Sigmanificient** [comment]
`97` items: `32.99%` done, `25.77%` in progress, `40` waiting

- [x] adafruit-nrfutil | #327199
- [x] cmake-lint | #327220
- [ ] mackup | #327240

### python3Packages

- [ ] actdiag | #330396
- [x] beaker | #327530
- [ ] biopandas
- [ ] blockdiag | #330396
- [ ] boto
- [x] braintree | #330290
- [ ] bx-python
- [ ] cassandra-driver
- [x] citeproc-py | #330237
- [x] cgroup-utils | #330287
- [ ] changefinder | #330311
- [ ] chart-studio | #330248
- [x] deform | #328917
- [ ] dm-control
- [ ] dodgy
- [ ] envs
- [x] expiringdict | #326665
- [ ] flask-restful
- [x] python311Packages.gocardless-pro | #326766
- [ ] hkdf
- [ ] http-ece
- [ ] jsonable
- [ ] libgpuarray
- [ ] locationsharinglib
- [ ] lockfile
- [ ] mohawk
- [ ] mwtypes
- [ ] mwxml
- [ ] neo
- [ ] nose
- [x] nose-pattern-exclude | #330439
- [ ] nose-timer
- [x] nose-warnings-filters | #330439
- [ ] nose-xunitmp
- [ ] nwdiag | #330396
- [ ] ofxhome | #327239
- [ ] ofxtools
- [ ] pastescript | #327246
- [ ] para | #330229
- [ ] paramz
- [ ] paver
- [ ] pid | #327372
- [ ] pprintpp
- [x] preggy | #329978 
- [ ] premailer
- [ ] prison
- [ ] prox-tv
- [ ] ptable
- [ ] pycdio
- [ ] pycontracts
- [ ] pydy
- [ ] pyexcel-ods
- [ ] pyexcel-xls
- [ ] pygatt
- [ ] pygeoip
- [x] pygogo | #329989
- [ ] pygtfs
- [ ] pylacrosse | #330500
- [x] pypass | #330222
- [x] pypillowfight | #327214
- [ ] pyquaternion | #330459
- [x] pyrfc3339 | #330496
- [ ] pysrt | #330441
- [x] pytest | #326541
- [ ] python-etcd | #330438
- [x] python-fedora | #327169
- [ ] python-hglib
- [ ] python-mapnik | #330432
- [x] pytimeparse | #330421
- [ ] pyutilib
- [ ] rauth | #330417
- [ ] safe | #330404
- [x] sampledata | #330029
- [ ] scikit-fuzzy
- [x] selectors2 | #330024
- [ ] seqdiag | #330396
- [x] simplebayes | #330018
- [x] spark-parser | #330001
- [x] sphinx-rtd-dark-mode | #330383
- [ ] sqlalchemy-mixins | #330493
- [x] sure | #330012
- [ ] svgutils | #330381
- [x] telfhash | #330005
- [ ] tempita | #330004
- [x] tissue | #330376
- [ ] traittypes | #330236
- [ ] trfl | #329976
- [x] unicode-slugify | #329998
- [x] vxi11 | #329994
- [ ] webhelpers | #330399
- [ ] weboob | #330436
- [x] xlwt | #330217
- [x] yanc | #329978
- [x] yarg | #327170

**Comment by emilazy** [comment]
> Maybe this will come in handy for stuff we need to patch: https://github.com/pytest-dev/nose2pytest.

I’ve packaged this in https://github.com/NixOS/nixpkgs/pull/330725 and am working on nose3 and its dependents.

**Comment by pyrox0** [comment]
I'm working on python-hglib, which has needed me to slightly redo the glean-sdk and glean-parser derivations. Should be submitting a PR as soon as nixpkgs-review finishes.

**Comment by pyrox0** [comment]
python3Packages.docker doesn't depend on nose, and hasn't for a while, should be good to check off.

**Comment by emilazy** [comment]
Ah, that’s my script being bad. It’s `docker-compose_1` in `pkgs/applications/virtualization/docker/compose_1.nix`. Which, uh, probably we could drop that?

**Comment by Sigmanificient** [comment]
`97` items: `45.36%` done, `16.49%` in progress, `37` waiting

- [x] adafruit-nrfutil | #327199
- [x] cmake-lint | #327220
- [x] mackup | #327240

### python3Packages

- [x] actdiag | #330396
- [x] beaker | #327530
- [x] biopandas | #330740
- [x] blockdiag | #330396
- [ ] boto
- [x] braintree | #330290
- [ ] bx-python | #330744
- [ ] cassandra-driver | #330760
- [x] citeproc-py | #330237
- [x] cgroup-utils | #330287
- [x] changefinder | #330311
- [ ] chart-studio | #330248
- [x] deform | #328917
- [ ] dm-control
- [x] dodgy | #330761
- [ ] envs | #330762
- [x] expiringdict | #326665
- [ ] flask-restful
- [x] gocardless-pro | #326766
- [x] habanero | #330768
- [x] hkdf | #330769
- [ ] http-ece | #330771
- [ ] isbnlib | #330773
- [ ] inotify | #330776
- [x] jsonable | #330777
- [ ] libgpuarray
- [ ] locationsharinglib | #330786
- [ ] lockfile
- [ ] mohawk
- [ ] mwtypes | #330805
- [ ] mwxml | #330805
- [ ] neo
- [ ] nose
- [ ] nose3 | #330741
- [x] nose-pattern-exclude | #330439
- [x] nose-timer | #330535
- [x] nose-warnings-filters | #330439
- [ ] nose-xunitmp
- [x] nwdiag | #330396
- [x] ofxhome | #327239
- [ ] ofxtools
- [x] pastescript | #327246
- [x] para | #330229
- [ ] paramz
- [ ] paver
- [ ] pid | #327372
- [ ] pprintpp | #330705
- [x] preggy | #329978 
- [ ] premailer | #330702
- [x] prison | #330700
- [ ] prox-tv | #330699
- [ ] ptable | #330696
- [ ] pycdio | #330695
- [ ] pycontracts | #330791
- [ ] pydy
- [x] pyexcel-ods | #330758
- [x] pyexcel-xls | #330758
- [ ] pygatt | #330688
- [ ] pygeoip | #330509
- [x] pygogo | #329989
- [x] pygtfs | #330504
- [x] pylacrosse | #330500
- [x] pypass | #330222
- [x] pypillowfight | #327214
- [x] pyquaternion | #330459
- [x] pyrfc3339 | #330496
- [x] pysrt | #330441
- [x] pytest | #326541
- [x] python-etcd | #330438
- [x] python-fedora | #327169
- [ ] python-hglib | #330739
- [x] python-mapnik | #330432
- [x] pytimeparse | #330421
- [x] pyutilib | #330759
- [x] rauth | #330417
- [ ] safe | #330404
- [x] sampledata | #330029
- [ ] scikit-fuzzy
- [x] selectors2 | #330024
- [x] seqdiag | #330396
- [x] simplebayes | #330018
- [x] spark-parser | #330001
- [x] sphinx-rtd-dark-mode | #330383
- [x] sqlalchemy-mixins | #330493
- [x] sure | #330012
- [ ] svgutils | #330381
- [x] telfhash | #330005
- [ ] tempita | #330004
- [x] tissue | #330376
- [x] traittypes | #330236
- [ ] trfl | #329976
- [x] unicode-slugify | #329998
- [x] vxi11 | #329994
- [x] webhelpers | #330399
- [x] weboob | #330436
- [x] xlwt | #330217
- [x] yanc | #329978
- [x] yarg | #327170

**Comment by lavafroth** [comment]
@emilazy python3Packages.mkl-service from the original list does not actually use nose. Guess we can get rid of it from the list.

**Comment by emilazy** [comment]
Looks like I just missed https://github.com/NixOS/nixpkgs/pull/324729. Fixed!

**Comment by Sigmanificient** [comment]
> [lockfile](http://git.openstack.org/cgit/openstack/pylockfile) is no longer maintained.


Apache airflow depends on it tho:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/apache-airflow/python-package.nix

**Comment by Sigmanificient** [comment]
What are its reverse dependencies?

**Comment by Sigmanificient** [comment]
https://github.com/pydy/pydy/issues/490

**Comment by Sigmanificient** [comment]
aws-google-auth may be dead: https://github.com/cevoaustralia/aws-google-auth/issues/284

**Comment by Sigmanificient** [comment]
@emilazy in your list, `candidates1` seems like a script error, i cant find it in nixpkgs

**Comment by emilazy** [comment]
Uh, yeah I have no idea where that came from.

> aws-google-auth may be dead: [cevoaustralia/aws-google-auth#284](https://github.com/cevoaustralia/aws-google-auth/issues/284)

Probably just drop it, although we’re not running its tests anyway.

**Comment by Sigmanificient** [comment]
`python3Packages.nbmerge` was done in the following pr:
- https://github.com/NixOS/nixpkgs/pull/324749

**Comment by Sigmanificient** [comment]
2 packages that have been already done @emilazy 

- #324749
- #324734

**Comment by emilazy** [comment]
24.05 uses Python 3.11 by default, so the packages here should mostly not be an issue. If you’re using Python 3.12, I would recommend using the unstable release until 24.11 is released.

(Also, most of these packages do already work with 3.12 on 24.11 as we have patched nose ourselves; we’re just trying to get them cleaned up.)

**Comment by mmorley0395** [comment]
Interesting-- I was using nix-shell and for some reason my nix-channel was set to unstable so it was pulling from there and therefore using 3.12 i guess. I don't need 3.12 specifically, so updated the channel and all is well now. Thanks for taking the time to explain!

**Comment by Sigmanificient** [comment]
Up to date list:

`114` items: `83.33%` done, `14.91%` in progress, `4` waiting

- [x] adafruit-nrfutil | #327199
- [x] aws-google-auth | #331352
- [ ] caffe
- [x] cmake-lint | #327220
- [x] gixy | #330741
- [x] grin | #330741
- [ ] hydrus
- [x] mackup | #327240
- [x] mbutil | #334492
- [ ] sasview
- [x] svtplay-dl | #330741
- [x] tarsnapper | #334462 
- [x] zk-shell | #333725

### python2Packages

- [ ] more-itertools

### python3Packages

- [x] actdiag | #330396
- [ ] apricot-select
- [x] androguard | #330535
- [x] beaker | #327530
- [x] biopandas | #330740
- [x] blockdiag | #330396
- [ ] blessings
- [ ] boto
- [x] braintree | #330290
- [x] bx-python | #330744
- [x] cassandra-driver | #330760
- [x] citeproc-py | #330237
- [x] cgroup-utils | #330287
- [x] changefinder | #330311
- [x] chart-studio | #330248
- [ ] clickhouse-driver | #330959
- [ ] clustershell
- [x] deform | #328917
- [x] dm-control | #330869
- [x] dodgy | #330761
- [ ] django-crontab
- [x] envs | #330762
- [x] expiringdict | #326665
- [x] flask-restful | #330876 
		‌‌‌‌‌‌‌ | #333579
- [x] gocardless-pro | #326766
- [x] habanero | #330768
- [ ] hkdf | #330769
- [x] http-ece | #330771
- [x] isbnlib | #330773
- [ ] inotify | #330776
- [x] jsonable | #330777
- [ ] jupiter-contrib-nbextensions
- [x] libgpuarray | #331841
- [x] locationsharinglib | #330786
- [ ] lockfile
- [ ] mhcflurry
- [x] mohawk | #331838
- [x] mongoengine | #334487
- [x] mwtypes | #330805
- [x] mwxml | #330805
- [x] neo | #330882
- [ ] nose
- [ ] nose2topytest
- [x] nose3 | #330741
- [x] nose-pattern-exclude | #330439
- [x] nose-timer | #330535
- [x] nose-warnings-filters | #330439
- [x] nose-xunitmp | #330869
- [x] nwdiag | #330396
- [x] ofxhome | #327239
- [x] ofxtools | #330866
- [x] pastescript | #327246
- [x] para | #330229
- [x] paramz | #330865
- [x] paver | #330858
- [ ] pid | #327372
- [x] pkutils | #330741
- [x] pprintpp | #330705
- [x] preggy | #329978 
- [ ] premailer | #330702
- [x] prison | #330700
- [x] prox-tv | #330699
- [x] ptable | #330696
- [x] pycdio | #330695
- [x] pycontracts | #330791
- [ ] pydy
- [x] pyexcel-ods | #330758
- [x] pyexcel-xls | #330758
- [x] pygatt | #330688
  | #332660
- [ ] pygeoip | #330509
- [x] pygogo | #329989
- [x] pygtfs | #330504
- [x] pylacrosse | #330500
- [x] pypass | #330222
- [x] pypillowfight | #327214
- [x] pyquaternion | #330459
- [x] pyrfc3339 | #330496
- [x] pysrt | #330441
- [x] pytest | #326541
- [x] python-etcd | #330438
- [x] python-fedora | #327169
- [x] python-hglib | #330739
- [x] python-mapnik | #330432
- [x] pytimeparse | #330421
- [x] pyutilib | #330759
- [x] rauth | #330417
- [ ] rx
- [x] safe | #330404 
- [x] sampledata | #330029
- [ ] scales | #334466
- [ ] scikit-fuzzy | #330854
- [x] selectors2 | #330024
- [x] seqdiag | #330396
- [ ] shap
- [x] simplebayes | #330018
- [x] spark-parser | #330001
- [x] sphinx-rtd-dark-mode | #330383
- [x] sqlalchemy-mixins | #330493
- [ ] stytra
- [x] sure | #330012
- [x] svgutils | #330381
- [x] telfhash | #330005
- [x] tempita | #330855
- [x] tissue | #330376
- [x] traittypes | #330236
- [ ] trfl | #329976
- [ ] uncompyle6
- [x] unicode-slugify | #329998
- [x] vxi11 | #329994
- [ ] webassets | #334483 
- [x] webhelpers | #330399
- [x] weboob | #330436
- [ ] wsme | #334472
- [x] xlwt | #330217
- [x] yanc | #329978
- [x] yarg | #327170

**Comment by Sigmanificient** [comment]
`umcompyle6` has `doCheck = pythonOlder "3.9";` but when using `nix-build -A python312.umcompyle6` tests are still being run, am I missing something @emilazy?

**Comment by Sigmanificient** [comment]
@emilazy should we drop `more-itertools`?

**Comment by SuperSandro2000** [comment]
> more-itertools

We can't, it is to widely used

**Comment by SuperSandro2000** [comment]
Try removing it and see what breaks. I would be in support to drop eg https://github.com/NixOS/nixpkgs/blob/21576aba43e604a95cc45c25248f39f3989ebe36/pkgs/servers/neard/default.nix

but we need to keep gimp intact https://github.com/NixOS/nixpkgs/blob/21576aba43e604a95cc45c25248f39f3989ebe36/pkgs/applications/graphics/gimp/default.nix#L137

I didn't find anything else uses python2 packages on the fast but I didn't search for all variants.

**Comment by emilazy** [comment]
Ignore Python 2; it is even more obsolete than nose. GIMP will be solved soon and I have plans for the rest of Python 2.

**Comment by emilazy** [comment]
Sorry for my long absence from working on this. I’ve cleaned up the list in the first comment to be easier to follow now that most users have been dealt with, and left a few reviews. I hope we can get this done before 24.11!

**Comment by emilazy** [comment]
PRs of my own in flight:

* https://github.com/NixOS/nixpkgs/pull/343776
* https://github.com/NixOS/nixpkgs/pull/343785
* https://github.com/NixOS/nixpkgs/pull/343797
* https://github.com/NixOS/nixpkgs/pull/343820

I’m looking at `python3Packages.mhcflurry` and `python3Packages.uncompyle6`, which are now almost the only remaining packages without an open PR.

**Comment by emilazy** [comment]
Turns out I was wrong about Python 2.7 packages not mattering… https://github.com/NixOS/nixpkgs/pull/343822

**Comment by emilazy** [comment]
@Sigmanificient Hi there, just checking if you plan to address the review comments on your pending PRs? No worries if not; I just want to make sure if you’d rather work more on them or would be happier for someone else to take over.

**Comment by emilazy** [comment]
We’ve finally reached the end of the road!

* https://github.com/NixOS/nixpkgs/pull/348699

Thanks again to everyone who helped out with this immense effort.

**Comment by Sigmanificient** [comment]
> @Sigmanificient Hi there, just checking if you plan to address the review comments on your pending PRs? No worries if not; I just want to make sure if you’d rather work more on them or would be happier for someone else to take over.

Hey, sorry i don't have a lot of avalaible time currently :<

**Comment by emilazy** [comment]
No worries! You played a huge part in this and I’m happy we’ve finally reached the finish line.
